### PR TITLE
Implement client-side routing with Page.js

### PR DESCRIPTION
**Pull request by flashbackzoo**
Routing required to implement React based CMS sections.

- Removes thirdparty dependency History.js
- Adds thirdparty dependency Page.js to manage client-side routing
- Adds a wrapper around Page.js for SilverStripe specific behaviour
- Increased minimum browser requirement to IE10. Native HTML History API routing requires IE10 or newer (necessitated by removal of History.js)
- PJAX pannel loading via now uses promises rather than callbacks
- Adds getClientConfig method to LeftAndMain which can be used to pass config to the client application.

@chillu

### CODE REVIEW

**Comment by chillu** [comment]
Note that the browser support requirement of IE10+ has been discussed and de-facto agreed on through https://groups.google.com/forum/#!topic/silverstripe-dev/ApaaWnlaHpM)

**Comment by chillu** [comment]
@silverstripe/core-team We're hoping to get this pull request over the line in the next 48h because we have an opportunity for staff to fix things during an internal hackday here at SilverStripe Ltd on NZ Friday, ideally without too much branching fragmentation. Can you please review in your respective areas of expertise (PHP and/or JS/React)? @flashbackzoo, @scott1702 and myself have largely peer-reviewed each others work here, but would be good to get the thumbs up from at least one other core committer since these pull requests lay the ground work for further React integration. You'll see some `@todo` tags, some of it is early stages, but we agree with @sminnee's guidance that we should try to integrate early and often in our product development cycle.

File: admin/code/LeftAndMain.php [hunk 1]
```diff
@@ -180,6 +180,50 @@ class LeftAndMain extends Controller implements PermissionProvider {
...
+		$combinedClientConfig = ['sections' => []];
+		$cmsClassNames = CMSMenu::get_cms_classes();
+
+		// Ordered by url_priority because the routing rules
+		// need to be registered in priority order on the client.
+		usort($cmsClassNames, function ($a, $b) {
+			$priorityA = Config::inst()->get($a, 'url_priority');
+			$priorityB = Config::inst()->get($b, 'url_priority');
+
+			if ($a == $b) {
```
Comment (kinglozzer) [inline admin/code/LeftAndMain.php]: Shouldn’t this be `if ($priorityA == $priorityB) {`?

Comment (flashbackzoo) [inline admin/code/LeftAndMain.php]: yeah it should

Comment (tractorcow) [inline admin/code/LeftAndMain.php]: These are already sorted in AdminRootController::rules() method. I would just move the logic from there to just before the `return` in `CMSMenu::get_cms_classes`.

**Comment by tractorcow** [comment]
Seems to have a bug trying to load the cms.

image

**Comment by tractorcow** [comment]
I've added another commit, fixing the issue with window.history.state being null on initial load, and putting some of the sorting for menu items into CMSMain class.

**Comment by flashbackzoo** [comment]
I think there could be something wrong with the bundles in that commit @tractorcow if you look at the lines added +38,671 LOC in `bundle-lib.js`

You probably just need to `npm run build` to minify it. Note you won't need to do this as soon :smile:

**Pull request merged**